crawl/crawl.go

```diff
@@ -1,178 +1,218 @@
+// been corrupted.
+func verify(logClient *client.LogClient, e entryData, rootHash []byte) error {
+	currentTreeSize := e.index + 1
+	pbh, err := logClient.GetProofByHash(context.Background(), e.merkleLeafHash, uint64(currentTreeSize))
+	if err != nil {
+		return fmt.Errorf("getting proof by hash: %w", err)
 	}
+
+	hasher := rfc6962.DefaultHasher
+	verifier := logverifier.New(hasher)
+	err = verifier.VerifyInclusionProof(e.index, currentTreeSize, pbh.AuditPath, rootHash, e.merkleLeafHash)
+	if err != nil {
+		var auditPathPrintable []string
+		for _, h := range pbh.AuditPath {
+			auditPathPrintable = append(auditPathPrintable, b64(h))
+		}
+		return fmt.Errorf("verify failed: VerifyInclusionProof(%d, %d, %s, %s, %s)=%w",
+			e.index, currentTreeSize, auditPathPrintable,
+			b64(rootHash), b64(e.merkleLeafHash), err)
+	}
+	return nil
 }
 
 func main() {
 	logURI := flag.String("log_uri", "https://oak.ct.letsencrypt.org/2022/", "CT log base URI")
-	batchSize := flag.Int("batch_size", 256, "Max number of entries to request at per call to get-entries")
+	batchSize := flag.Int("batch_size", 256, "Max number of entries to request per call to get-entries")
 	numWorkers := flag.Int("num_workers", 2, "Number of concurrent workers")
-	startIndex := flag.Int64("start_index", 0, "Log index to start scanning at")
-	endIndex := flag.Int64("end_index", 0, "Log index to end scanning at (non-inclusive, 0 = end of log)")
+	flag.Usage = func() {
+		fmt.Print(`This tool gets the current STH of a log, then fetches all
+entries up to that STH's tree_size, building the root hash as it goes. If the
+root hash doesn't match, it exits with an error.
+
+It also checks get-proof-by-hash along the way, builds the implied root hash
+from the response, and checks that it matches the currently calculated root.
+If there's a mismatch, this tool exits with an error.
+
+`)
+		flag.PrintDefaults()
+	}
 	flag.Parse()
 
+	if *logURI == "" {
+		log.Fatal("must provide the -log_uri flag")
+	}
+
 	logClient, err := client.New(*logURI, &http.Client{
 		Timeout: 10 * time.Second,
-		Transport: &http.Transport{
-			TLSHandshakeTimeout:   30 * time.Second,
-			ResponseHeaderTimeout: 30 * time.Second,
-			MaxIdleConnsPerHost:   10,
-			DisableKeepAlives:     false,
-			MaxIdleConns:          100,
-			IdleConnTimeout:       90 * time.Second,
-			ExpectContinueTimeout: 1 * time.Second,
-		},
 	}, jsonclient.Options{UserAgent: "le-ct-crawler/0.1"})
 	if err != nil {
-		log.Fatal("Failed to create log client")
+		log.Fatalf("creating log client: %s", err)
 	}
 
+	// Fetch whatever the log is serving as the current STH. Currently we just
+	// trust the contents; we don't verify the signature.
+	sth, err := logClient.GetSTH(context.Background())
+	if err != nil {
+		log.Fatalf("getting STH: %s", err)
+	}
+
+	log.Printf("verifying tree integrity at size %d (%s) with root hash %s", sth.TreeSize,
+		time.Unix(0, int64(sth.Timestamp*1e6)), sth.SHA256RootHash.Base64String())
```

Comment:
You can just use `time.UnixMilli` instead of converting to nanoseconds:
```suggestion
		time.UnixMilli(int64(sth.Timestamp)), sth.SHA256RootHash.Base64String())
```